lib/goliath/request.rb

```diff
@@ -197,29 +197,46 @@ def post_process(results)
         callback do
           begin
             @response.status, @response.headers, @response.body = status, headers, body
-            @response.each { |chunk| @conn.send_data(chunk) }
-
-            elapsed_time = (Time.now.to_f - @env[:start_time]) * 1000
-            begin
-              Goliath::Request.log_block.call(@env, @response, elapsed_time)
-            rescue => err
-              # prevent an infinite loop if the block raised an error
-              @env[RACK_LOGGER].error("log block raised #{err}")
-            end
 
-            @conn.terminate_request(keep_alive)
+            stream_data(@response.each) do
+              terminate_request
+            end
           rescue Exception => e
             server_exception(e)
           end
         end
-
       rescue Exception => e
         server_exception(e)
       end
     end
 
     private
 
+    # Writes each chunk of the response data in a new tick. This achieves
```

Comment:
> This is inefficient. First, it leads to high memory consumption on large
response bodies, because the content is buffered in-memory before it is
sent to the client. Second, if the response body takes time to generate,
the client needs to wait until the end before it can start downloading
the response body.

As a sanity check: if the client generates a single large response, the code below doesn't help much -- right? All we've done is split delivery across multiple ticks (does EM "batch" these ticks -- been a while since I've looked at the internals). The case where this change _will_ help is if the client is generating a large streaming response -- correct?

Reply:
Correct 👌